File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
private:
    bool solve(vector<int> &arr, vector<bool> &visited, int i){
        if(i >=arr.size() || i < 0) return false;
        if(visited[i] == true) return false;
        
        // target case
        if(arr[i] == 0){
            // possible[i] = 1;
            return true;
        }

        visited[i] = true;
        // two ways: 
            // to go positive side
            // to go negative side
        
        int rightIdx = arr[i] + i;
        int leftIdx = i - arr[i];
        
        bool rightPossible = solve(arr, visited, rightIdx);
        if(rightPossible) return true;
        bool leftPossible = solve(arr, visited, leftIdx);
        if(leftPossible || rightPossible) return true;
        return false;
    }


public:
    bool canReach(vector<int>& arr, int start) {
        int n = arr.size();
        vector<bool> visited(n, false);

        // visited[start] = true;
        return solve(arr, visited, possible, start);

    }
};

File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution:     
    def canReach(self, arr: List[int], start: int) -> bool:
        visited = [False] * len(arr)

        def solve(idx: int ) -> bool:
            if idx  >= len(arr) or idx < 0:
                return False
            if visited[idx] == True:
                return False
            
            if arr[idx] == 0:
                return True
            
            visited[idx] = True

            right_idx = idx + arr[idx]
            left_idx = idx - arr[idx]

            rightPossible = solve(right_idx)
            leftPossible = solve(left_idx)

            if rightPossible == True or leftPossible == True:
                return True
            return False

        return solve(start)
        